wolk/service/firmware_update/FirmwareUpdateService.cpp

```diff
@@ -0,0 +1,292 @@
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Check if there's already an installation session ongoing
+    if (FileSystemUtils::isFilePresent(m_sessionFile))
+    {
+        LOG(WARN) << "Received 'FirmwareUpdateInstallMessage' but an installation is already ongoing.";
+        return;
+    }
+
+    // Check with the installer
+    auto status = m_firmwareInstaller->installFirmware(message.getFile());
+    switch (status)
+    {
+    case InstallResponse::FAILED_TO_INSTALL:
+        sendStatusMessage(FirmwareUpdateStatus::ERROR, FirmwareUpdateError::INSTALLATION_FAILED);
+        return;
+    case InstallResponse::NO_FILE:
+        sendStatusMessage(FirmwareUpdateStatus::ERROR, FirmwareUpdateError::UNKNOWN_FILE);
+        return;
+    case InstallResponse::WILL_INSTALL:
+        if (!storeSessionFile())
+        {
+            LOG(ERROR) << "Failed to store session file.";
+            sendStatusMessage(FirmwareUpdateStatus::ERROR, FirmwareUpdateError::UNKNOWN);
+            return;
+        }
+        sendStatusMessage(FirmwareUpdateStatus::INSTALLING);
+        return;
+    case InstallResponse::INSTALLED:
+        sendStatusMessage(FirmwareUpdateStatus::SUCCESS);
+        break;
+    }
+}
+
+void FirmwareUpdateService::onFirmwareAbort(const std::string& /** deviceKey **/,
+                                            const FirmwareUpdateAbortMessage& /** message **/)
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Check if the session file is present
+    if (FileSystemUtils::isFilePresent(m_sessionFile))
+        m_firmwareInstaller->abortFirmwareInstall();
+}
+
+void FirmwareUpdateService::sendStatusMessage(FirmwareUpdateStatus status, FirmwareUpdateError error)
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Create the status message
+    auto statusMessage = FirmwareUpdateStatusMessage(status, error);
+    auto message = std::shared_ptr<Message>(m_protocol.makeOutboundMessage(m_deviceKey, statusMessage));
+    if (message == nullptr)
+    {
+        LOG(ERROR) << "Failed to generate outbound FirmwareUpdateStatusMessage.";
+        return;
+    }
+    m_connectivityService.publish(message);
+}
+
+void FirmwareUpdateService::queueStatusMessage(FirmwareUpdateStatus status, FirmwareUpdateError error)
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Create the status message
+    auto statusMessage = FirmwareUpdateStatusMessage(status, error);
+    auto message = std::shared_ptr<Message>(m_protocol.makeOutboundMessage(m_deviceKey, statusMessage));
+    if (message == nullptr)
+    {
+        LOG(ERROR) << "Failed to generate outbound FirmwareUpdateStatusMessage.";
+        return;
+    }
+    m_queue.push(message);
+}
+
+bool FirmwareUpdateService::storeSessionFile()
+{
+    return FileSystemUtils::createFileWithContent(m_sessionFile, m_firmwareInstaller->getFirmwareVersion());
+}
+
+void FirmwareUpdateService::deleteSessionFile()
+{
+    FileSystemUtils::deleteFile(m_sessionFile);
+}
+
+void FirmwareUpdateService::reportParameters()
```

Comment:
Okay, so, you just mind that there is `m_dataService.publishParameters()`. Also, you can't just `return parameters when requested`, since there is no kind of mechanism like that, all parameters are collected and published whenever the device wishes to. I'll remove the `publishParameters()` call, and they will travel out with the next call of `wolk->publish()` then, the call of which does not have to be guaranteed.

If you would want me to implement an alternative service (or service state) where Firmware Update is disabled, I could do that. Thought it was out of scope for this rework.